### PR TITLE
feat: rewrite AmountDeriver in Vyper

### PR DESCRIPTION
**Pull request by 0xbok**

### CODE REVIEW

**Review by 0age — APPROVED** [review]

File: vyper/contracts/libs/AmountDeriver.vy [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    # Multiply the numerator by the value and ensure no overflow occurs.
+    valueTimesNumerator: uint256 = val * numerator
+
+    # Divide that value by the denominator to get the new value.
+    newValue: uint256 = valueTimesNumerator / denominator
+
+    # Ensure that division gave a final result with no remainder.
+    exact: bool = ((newValue * denominator) / numerator) == val
+    if not exact:
+        raise "inexact fraction"
```
[the remark this answers is not in the document]
Comment (charles-cooper) [reply vyper/contracts/libs/AmountDeriver.vy:96]: or, inline the whole thing -- `assert ((newValue * denominator) / numerator) == val`

[the remark this answers is not in the document]
Comment (0xbok) [reply vyper/contracts/libs/AmountDeriver.vy]: inlined.

File: vyper/contracts/libs/AmountDeriver.vy [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+            extraCeiling)
+
+        # Division is performed without zero check as it cannot be zero.
+        newAmount: uint256 = totalBeforeDivision / duration
+
+        # Return the current amount (expressed as endAmount internally).
+        return newAmount
+
+    # Return the original amount (now expressed as endAmount internally).
+    return endAmount
```
Comment (charles-cooper) [inline vyper/contracts/libs/AmountDeriver.vy:63]: nit: stylistically, might be good to fast path out `if startAmount == endAmount`. then the more complicated part can be less indented.

**Review by charles-cooper — APPROVED** [review]

File: vyper/contracts/libs/AmountDeriver.vy (hunk 2 above)
Comment (0age) [reply vyper/contracts/libs/AmountDeriver.vy]: I also think `# Division is performed without zero check as it cannot be zero.` comment is maybe not true? Not sure whether vyper does a zero check on division, guessing it does though

Comment (charles-cooper) [reply vyper/contracts/libs/AmountDeriver.vy]: > I also think `# Division is performed without zero check as it cannot be zero.` comment is maybe not true? Not sure whether vyper does a zero check on division, guessing it does though

vyper division does do a 0 check. if you wanted to really skip the zero check, could use `unsafe_div(a, b)`

Comment (0xbok) [reply vyper/contracts/libs/AmountDeriver.vy]: I need to be more careful with the comments than just copying from the reference :) 
Will go with `unsafe_div`.

**Pull request merged**